src/jit/morph.cpp

```diff
@@ -19624,3 +19534,136 @@ bool Compiler::fgMorphCombineSIMDFieldAssignments(BasicBlock* block, GenTreePtr
+            if (!isSideEffectFree && (popExpr->OperGet() == GT_COMMA))
+            {
+                isSideEffectFree = ((popExpr->gtGetOp1()->gtFlags & GTF_ALL_EFFECT) == 0) &&
+                                   ((popExpr->gtGetOp2()->gtFlags & GTF_ALL_EFFECT) == 0);
+            }
+            noway_assert(isSideEffectFree);
+
+            nextMorphStmt = popStmt->gtNextStmt;
+        }
+
+        // Next skip any GT_NOP nodes after the pop
+        nextMorphStmt = skipNopStmts(nextMorphStmt);
+    }
+#endif // !FEATURE_CORECLR && _TARGET_AMD64_
+
+    // Check that the rest stmts in the block are expected.
+    if (nextMorphStmt != nullptr)
+    {
+        GenTree* callExpr = callStmt->gtStmtExpr;
+        if (callExpr->gtOper != GT_ASG)
+        {
+            // The next stmt must be GT_RETURN(TYP_VOID).
+            noway_assert(info.compRetType == TYP_VOID || info.compRetType == TYP_STRUCT);
+            GenTreeStmt* retStmt = nextMorphStmt;
+            GenTreePtr   retExpr = retStmt->gtStmtExpr;
+            noway_assert(retExpr->gtOper == GT_RETURN);
+
+            nextMorphStmt = retStmt->gtNextStmt;
+        }
+        else
+        {
+            noway_assert(callExpr->gtGetOp1()->OperIsLocal());
+            unsigned callResultLclNumber = callExpr->gtGetOp1()->AsLclVarCommon()->gtLclNum;
+
+#if defined(FEATURE_CORECLR)
+
+            // We can have a move from the call result to an lvaInlineeReturnSpillTemp.
+            // However, we can't check that this assignment was created there.
+            if (nextMorphStmt->gtStmtExpr->gtOper == GT_ASG)
+            {
+                GenTreeStmt* moveStmt = nextMorphStmt;
+                GenTree*     moveExpr = nextMorphStmt->gtStmtExpr;
+                noway_assert(moveExpr->gtGetOp1()->OperIsLocal() && moveExpr->gtGetOp2()->OperIsLocal());
+
+                unsigned srcLclNum = moveExpr->gtGetOp2()->AsLclVarCommon()->gtLclNum;
+                noway_assert(srcLclNum == callResultLclNumber);
+                unsigned dstLclNum  = moveExpr->gtGetOp1()->AsLclVarCommon()->gtLclNum;
+                callResultLclNumber = dstLclNum;
+
+                nextMorphStmt = moveStmt->gtNextStmt;
+            }
+            if (nextMorphStmt != nullptr)
+#endif
+            {
+                GenTreeStmt* retStmt = nextMorphStmt;
+                GenTreePtr   retExpr = nextMorphStmt->gtStmtExpr;
+                noway_assert(retExpr->gtOper == GT_RETURN);
+
+                GenTreePtr treeWithLcl = retExpr->gtGetOp1();
+                while (treeWithLcl->gtOper == GT_CAST)
+                {
+                    noway_assert(!treeWithLcl->gtOverflow());
+                    treeWithLcl = treeWithLcl->gtGetOp1();
+                }
+
+                noway_assert(callResultLclNumber == treeWithLcl->AsLclVarCommon()->gtLclNum);
+
+                nextMorphStmt = retStmt->gtNextStmt;
+            }
+        }
+    }
+    return nextMorphStmt == nullptr;
```

Comment:
Maybe we should do `assert(nextMorphStmt == nullptr)` here, to catch possible regressions in the future.